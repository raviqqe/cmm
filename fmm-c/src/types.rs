use crate::names::*;
use fmm::types::{self, Type};
use std::collections::FnvHashMap;

pub fn compile_typed_name(type_: &Type, name: &str, type_ids: &FnvHashMap<Type, String>) -> String {
    match type_ {
        Type::Function(function) => {
            compile_function_name(function, &format!("(*{})", name), type_ids)
        }
        Type::Primitive(primitive) => compile_primitive_type_id(*primitive) + " " + name,
        Type::Record(record) => compile_record_type_id(record, type_ids) + " " + name,
        Type::Pointer(pointer) => {
            compile_typed_name(pointer.element(), &format!("*{}", name), type_ids)
        }
        Type::Union(union) => compile_union_type_id(union, type_ids) + " " + name,
    }
}

pub fn compile_function_name(
    function: &types::Function,
    name: &str,
    type_ids: &FnvHashMap<Type, String>,
) -> String {
    compile_typed_name(
        function.result(),
        &format!(
            "{}({})",
            name,
            function
                .arguments()
                .iter()
                .enumerate()
                .map(|(index, type_)| compile_typed_name(
                    type_,
                    &generate_argument_name(index),
                    type_ids
                ))
                .collect::<Vec<_>>()
                .join(",")
        ),
        type_ids,
    )
}

pub fn compile_type_id(type_: &Type, type_ids: &FnvHashMap<Type, String>) -> String {
    match type_ {
        Type::Function(function) => compile_function_name(function, "(*)", type_ids),
        Type::Primitive(primitive) => compile_primitive_type_id(*primitive),
        Type::Record(record) => compile_record_type_id(record, type_ids),
        Type::Pointer(pointer) => compile_typed_name(pointer.element(), "*", type_ids),
        Type::Union(union) => compile_union_type_id(union, type_ids),
    }
}

pub fn compile_atomic_pointer_type_id(type_: &Type, type_ids: &FnvHashMap<Type, String>) -> String {
    match type_ {
        Type::Function(function) => compile_function_name(function, "(*_Atomic *)", type_ids),
        Type::Pointer(pointer) => compile_typed_name(pointer.element(), "_Atomic *", type_ids),
        Type::Primitive(_) | Type::Record(_) | Type::Union(_) => {
            "_Atomic ".to_owned() + &compile_type_id(type_, type_ids) + " *"
        }
    }
}

pub fn compile_primitive_type_id(primitive: types::Primitive) -> String {
    match primitive {
        types::Primitive::Boolean => "bool",
        types::Primitive::Float32 => "float",
        types::Primitive::Float64 => "double",
        types::Primitive::Integer8 => "uint8_t",
        types::Primitive::Integer32 => "uint32_t",
        types::Primitive::Integer64 => "uint64_t",
        types::Primitive::PointerInteger => "size_t",
    }
    .into()
}

pub fn compile_record_type_id(record: &types::Record, type_ids: &FnvHashMap<Type, String>) -> String {
    "struct ".to_owned() + &type_ids[&record.clone().into()]
}

pub fn compile_union_type_id(union: &types::Union, type_ids: &FnvHashMap<Type, String>) -> String {
    "union ".to_owned() + &type_ids[&union.clone().into()]
}

pub fn compile_record_fields(record: &types::Record, type_ids: &FnvHashMap<Type, String>) -> String {
    record
        .fields()
        .iter()
        .enumerate()
        .map(|(index, type_)| {
            compile_typed_name(type_, &generate_record_field_name(index), type_ids) + ";"
        })
        .collect::<Vec<_>>()
        .join("")
}

pub fn compile_union_members(union: &types::Union, type_ids: &FnvHashMap<Type, String>) -> String {
    union
        .members()
        .iter()
        .enumerate()
        .map(|(index, type_)| {
            compile_typed_name(type_, &generate_union_member_name(index), type_ids) + ";"
        })
        .collect::<Vec<_>>()
        .join("")
}
