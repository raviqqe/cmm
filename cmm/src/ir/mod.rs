mod allocate_heap;
mod allocate_stack;
mod alternative;
mod argument;
mod arithmetic_operation;
mod assignment;
mod atomic_load;
mod atomic_store;
mod bitcast;
mod call;
mod compare_and_swap;
mod comparison_operation;
mod deconstruct_record;
mod default_alternative;
mod expression;
mod function_declaration;
mod function_definition;
mod if_;
mod instruction;
mod load;
mod module;
mod pointer_address;
mod primitive;
mod record;
mod record_address;
mod return_;
mod store;
mod switch;
mod union;
mod union_address;
mod variable;
mod variable_declaration;
mod variable_definition;

pub use allocate_heap::*;
pub use allocate_stack::*;
pub use alternative::*;
pub use argument::*;
pub use arithmetic_operation::*;
pub use assignment::*;
pub use atomic_load::*;
pub use atomic_store::*;
pub use bitcast::*;
pub use call::*;
pub use compare_and_swap::*;
pub use comparison_operation::*;
pub use deconstruct_record::*;
pub use default_alternative::*;
pub use expression::*;
pub use function_declaration::*;
pub use function_definition::*;
pub use if_::*;
pub use instruction::*;
pub use load::*;
pub use module::*;
pub use pointer_address::*;
pub use primitive::*;
pub use record::*;
pub use record_address::*;
pub use return_::*;
pub use store::*;
pub use switch::*;
pub use union::*;
pub use union_address::*;
pub use variable::*;
pub use variable_declaration::*;
pub use variable_definition::*;
