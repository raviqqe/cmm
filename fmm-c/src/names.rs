pub fn generate_argument_name(index: usize) -> String {
    format!("a{}", index)
}

pub fn generate_record_field_name(index: usize) -> String {
    format!("e{}", index)
}

pub fn generate_union_member_name(index: usize) -> String {
    format!("m{}", index)
}

pub fn generate_record_type_name(index: usize) -> String {
    format!("r{}", index)
}

pub fn generate_union_type_name(index: usize) -> String {
    format!("u{}", index)
}
