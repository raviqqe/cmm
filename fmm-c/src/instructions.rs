use fmm::ir::*;

pub fn compile_block(_block: &Block) -> String {
    todo!()
}
