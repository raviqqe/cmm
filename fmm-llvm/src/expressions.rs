use crate::union::compile_union_cast;

use super::types::*;
use fmm::ir::*;
use fmm::types;
use inkwell::values::BasicValue;
use std::collections::HashMap;

pub fn compile_expression<'c>(
    builder: &inkwell::builder::Builder<'c>,
    expression: &Expression,
    variables: &HashMap<String, inkwell::values::BasicValueEnum<'c>>,
    context: &'c inkwell::context::Context,
    target_data: &inkwell::targets::TargetData,
) -> inkwell::values::BasicValueEnum<'c> {
    let compile_expression = |expression: &Expression| {
        compile_expression(builder, expression, variables, context, target_data)
    };

    match expression {
        Expression::AlignOf(align_of) => compile_align_of(align_of, context, target_data).into(),
        Expression::BitCast(bit_cast) => {
            compile_bit_cast(builder, bit_cast, context, target_data, &compile_expression)
        }
        Expression::BitwiseOperation(operation) => {
            compile_bitwise_operation(builder, operation, &compile_expression).into()
        }
        Expression::Primitive(primitive) => compile_primitive(*primitive, context, target_data),
        Expression::Record(record) => {
            let mut value = compile_record_type(record.type_(), context, target_data).const_zero();

            for (index, element) in record.elements().iter().enumerate() {
                value = builder
                    .build_insert_value(value, compile_expression(element), index as u32, "")
                    .unwrap()
                    .into_struct_value();
            }

            value.into()
        }
        Expression::SizeOf(size_of) => compile_size_of(size_of, context, target_data).into(),
        Expression::Undefined(undefined) => compile_undefined(undefined, context, target_data),
        Expression::Union(union) => {
            let member = compile_expression(union.member());

            let value = context.const_struct(
                &[
                    member.get_type().const_zero(),
                    compile_union_member_padding_type(
                        union.type_(),
                        union.member_index(),
                        context,
                        target_data,
                    )
                    .const_zero()
                    .into(),
                ],
                false,
            );

            compile_union_cast(
                builder,
                builder
                    .build_insert_value(value, member, 0, "")
                    .unwrap()
                    .as_basic_value_enum(),
                compile_union_type(union.type_(), context, target_data).into(),
            )
        }
        Expression::Variable(variable) => variables[variable.name()],
    }
}

pub fn compile_constant_expression<'c>(
    expression: &Expression,
    variables: &HashMap<String, inkwell::values::BasicValueEnum<'c>>,
    context: &'c inkwell::context::Context,
    target_data: &inkwell::targets::TargetData,
) -> inkwell::values::BasicValueEnum<'c> {
    let compile_expression = |expression: &Expression| {
        compile_constant_expression(expression, variables, context, target_data)
    };

    match expression {
        Expression::AlignOf(align_of) => compile_align_of(align_of, context, target_data).into(),
        Expression::BitCast(bit_cast) => compile_bit_cast(
            &context.create_builder(),
            bit_cast,
            context,
            target_data,
            &compile_expression,
        ),
        Expression::BitwiseOperation(operation) => {
            compile_bitwise_operation(&context.create_builder(), operation, &compile_expression)
                .into()
        }
        Expression::Primitive(primitive) => compile_primitive(*primitive, context, target_data),
        Expression::Record(record) => context
            .const_struct(
                &record
                    .elements()
                    .iter()
                    .map(|expression| compile_expression(expression))
                    .collect::<Vec<_>>(),
                false,
            )
            .into(),
        Expression::SizeOf(size_of) => compile_size_of(size_of, context, target_data).into(),
        Expression::Undefined(undefined) => compile_undefined(undefined, context, target_data),
        Expression::Union(union) => context
            .const_struct(
                &[
                    compile_expression(union.member()),
                    compile_union_member_padding_type(
                        union.type_(),
                        union.member_index(),
                        context,
                        target_data,
                    )
                    .const_zero()
                    .into(),
                ],
                false,
            )
            .into(),
        Expression::Variable(variable) => variables[variable.name()],
    }
}

fn compile_align_of<'c>(
    align_of: &AlignOf,
    context: &'c inkwell::context::Context,
    target_data: &inkwell::targets::TargetData,
) -> inkwell::values::IntValue<'c> {
    compile_pointer_integer(
        target_data.get_abi_alignment(&compile_type(align_of.type_(), context, target_data)) as u64,
        context,
        target_data,
    )
}

fn compile_bit_cast<'c>(
    builder: &inkwell::builder::Builder<'c>,
    bit_cast: &BitCast,
    context: &'c inkwell::context::Context,
    target_data: &inkwell::targets::TargetData,
    compile_expression: &impl Fn(&Expression) -> inkwell::values::BasicValueEnum<'c>,
) -> inkwell::values::BasicValueEnum<'c> {
    builder.build_bitcast(
        compile_expression(bit_cast.expression()),
        compile_type(bit_cast.to(), context, target_data),
        "",
    )
}

fn compile_bitwise_operation<'c>(
    builder: &inkwell::builder::Builder<'c>,
    operation: &BitwiseOperation,
    compile_expression: &impl Fn(&Expression) -> inkwell::values::BasicValueEnum<'c>,
) -> inkwell::values::IntValue<'c> {
    let lhs = compile_expression(operation.lhs()).into_int_value();
    let rhs = compile_expression(operation.rhs()).into_int_value();

    match operation.operator() {
        fmm::ir::BitwiseOperator::And => builder.build_and(lhs, rhs, ""),
        fmm::ir::BitwiseOperator::Or => builder.build_or(lhs, rhs, ""),
    }
}

fn compile_size_of<'c>(
    size_of: &SizeOf,
    context: &'c inkwell::context::Context,
    target_data: &inkwell::targets::TargetData,
) -> inkwell::values::IntValue<'c> {
    compile_pointer_integer(
        target_data.get_store_size(&compile_type(size_of.type_(), context, target_data)) as u64,
        context,
        target_data,
    )
}

fn compile_undefined<'c>(
    undefined: &Undefined,
    context: &'c inkwell::context::Context,
    target_data: &inkwell::targets::TargetData,
) -> inkwell::values::BasicValueEnum<'c> {
    match undefined.type_() {
        types::Type::Function(function) => {
            compile_function_pointer_type(function, context, target_data)
                .const_zero()
                .into()
        }
        types::Type::Primitive(primitive) => {
            compile_undefined_primitive(*primitive, context, target_data)
        }
        types::Type::Pointer(pointer) => compile_pointer_type(pointer, context, target_data)
            .const_zero()
            .into(),
        types::Type::Record(record) => compile_record_type(record, context, target_data)
            .const_zero()
            .into(),
        types::Type::Union(union) => compile_union_type(union, context, target_data)
            .const_zero()
            .into(),
    }
}

// TODO Is signed options for the const_int method correct?
fn compile_primitive<'c>(
    primitive: Primitive,
    context: &'c inkwell::context::Context,
    target_data: &inkwell::targets::TargetData,
) -> inkwell::values::BasicValueEnum<'c> {
    match primitive {
        Primitive::Boolean(boolean) => context.bool_type().const_int(boolean as u64, false).into(),
        Primitive::Float32(number) => context.f32_type().const_float(number as f64).into(),
        Primitive::Float64(number) => context.f64_type().const_float(number as f64).into(),
        Primitive::Integer8(number) => context.i8_type().const_int(number as u64, false).into(),
        Primitive::Integer32(number) => context.i32_type().const_int(number as u64, false).into(),
        Primitive::Integer64(number) => context.i64_type().const_int(number, false).into(),
        Primitive::PointerInteger(number) => {
            compile_pointer_integer(number as u64, context, target_data).into()
        }
    }
}

pub fn compile_pointer_integer<'c>(
    number: u64,
    context: &'c inkwell::context::Context,
    target_data: &inkwell::targets::TargetData,
) -> inkwell::values::IntValue<'c> {
    compile_pointer_integer_type(context, target_data).const_int(number, false)
}

// TODO Refactor this by matching with types::Primitive directly.
fn compile_undefined_primitive<'c>(
    type_: types::Primitive,
    context: &'c inkwell::context::Context,
    target_data: &inkwell::targets::TargetData,
) -> inkwell::values::BasicValueEnum<'c> {
    match compile_primitive_type(type_, context, target_data) {
        inkwell::types::BasicTypeEnum::FloatType(float) => float.const_zero().into(),
        inkwell::types::BasicTypeEnum::IntType(integer) => integer.const_zero().into(),
        inkwell::types::BasicTypeEnum::PointerType(pointer) => pointer.const_zero().into(),
        _ => unreachable!(),
    }
}
