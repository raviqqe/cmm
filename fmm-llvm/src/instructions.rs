use crate::{
    calling_convention::compile_calling_convention, error::CompileError, expressions::*,
    instruction_configuration::InstructionFunctionSet, types, union::compile_union_cast,
};
use fmm::ir::*;
use inkwell::values::BasicValue;
use std::convert::TryFrom;

pub fn compile_block<'c>(
    builder: &inkwell::builder::Builder<'c>,
    block: &Block,
    destination: Option<inkwell::basic_block::BasicBlock<'c>>,
    variables: &hamt::Map<String, inkwell::values::BasicValueEnum<'c>>,
    context: &'c inkwell::context::Context,
    target_data: &inkwell::targets::TargetData,
    instruction_function_set: &InstructionFunctionSet<'c>,
) -> Result<Option<inkwell::values::BasicValueEnum<'c>>, CompileError> {
    let mut variables = variables.clone();

    for instruction in block.instructions() {
        let value = compile_instruction(
            builder,
            instruction,
            &variables,
            context,
            target_data,
            instruction_function_set,
        )?;

        if let Some(value) = value {
            if let Some((name, _)) = instruction.value() {
                variables = variables.insert(name.into(), value);
            }
        }
    }

    Ok(compile_terminal_instruction(
        builder,
        block.terminal_instruction(),
        destination,
        &variables,
        context,
        target_data,
        instruction_function_set,
    ))
}

fn compile_instruction<'c>(
    builder: &inkwell::builder::Builder<'c>,
    instruction: &Instruction,
    variables: &hamt::Map<String, inkwell::values::BasicValueEnum<'c>>,
    context: &'c inkwell::context::Context,
    target_data: &inkwell::targets::TargetData,
    instruction_function_set: &InstructionFunctionSet<'c>,
) -> Result<Option<inkwell::values::BasicValueEnum<'c>>, CompileError> {
    let compile_expression =
        |expression| compile_expression(builder, expression, variables, context, target_data);
    let compile_type = |type_| types::compile(type_, context, target_data);

    Ok(match instruction {
        Instruction::AllocateHeap(allocate) => Some(
            builder
                .build_call(
                    instruction_function_set.allocate_function,
                    &[compile_expression(allocate.size()).into()],
                    allocate.name(),
                )
                .try_as_basic_value()
                .left()
                .unwrap(),
        ),
        Instruction::AllocateStack(allocate) => Some(
            builder
                .build_alloca(compile_type(allocate.type_()), allocate.name())
                .into(),
        ),
        Instruction::AtomicLoad(load) => {
            let value = builder.build_load(
                compile_expression(load.pointer()).into_pointer_value(),
                load.name(),
            );

            value
                .as_instruction_value()
                .unwrap()
                .set_atomic_ordering(compile_atomic_ordering(load.ordering()))?;

            Some(value)
        }
        Instruction::AtomicOperation(operation) => Some(
            builder
                .build_atomicrmw(
                    match operation.operator() {
                        AtomicOperator::Add => inkwell::AtomicRMWBinOp::Add,
                        AtomicOperator::Subtract => inkwell::AtomicRMWBinOp::Sub,
                    },
                    compile_expression(operation.pointer()).into_pointer_value(),
                    compile_expression(operation.value()).into_int_value(),
                    compile_atomic_ordering(operation.ordering()),
                )?
                .into(),
        ),
        Instruction::AtomicStore(store) => {
            let value = builder.build_store(
                compile_expression(store.pointer()).into_pointer_value(),
                compile_expression(store.value()),
            );

            value.set_atomic_ordering(compile_atomic_ordering(store.ordering()))?;

            None
        }
        Instruction::Call(call) => {
            let value = builder.build_call(
                inkwell::values::CallableValue::try_from(
                    compile_expression(call.function()).into_pointer_value(),
                )
                .unwrap(),
                &call
                    .arguments()
                    .iter()
                    .map(|expression| compile_expression(expression).into())
                    .collect::<Vec<_>>(),
                call.name(),
            );

            value.set_tail_call(true);
            value.set_call_convention(compile_calling_convention(
                call.type_().calling_convention(),
            )?);

            Some(value.try_as_basic_value().left().unwrap())
        }
        Instruction::CompareAndSwap(cas) => Some(
            builder
                .build_extract_value(
                    builder.build_cmpxchg(
                        compile_expression(cas.pointer()).into_pointer_value(),
                        compile_expression(cas.old_value()),
                        compile_expression(cas.new_value()),
                        compile_atomic_ordering(cas.success_ordering()),
                        compile_atomic_ordering(cas.failure_ordering()),
                    )?,
                    1,
                    cas.name(),
                )
                .unwrap(),
        ),
        Instruction::DeconstructRecord(deconstruct) => builder.build_extract_value(
            compile_expression(deconstruct.record()).into_struct_value(),
            deconstruct.field_index() as u32,
            deconstruct.name(),
        ),
        Instruction::DeconstructUnion(deconstruct) => builder.build_extract_value(
            compile_union_cast(
                builder,
                compile_expression(deconstruct.union()),
                types::compile_union_member(
                    deconstruct.type_(),
                    deconstruct.member_index(),
                    context,
                    target_data,
                )
                .into(),
            )
            .into_struct_value(),
            0,
            deconstruct.name(),
        ),
        Instruction::Fence(fence) => {
            builder.build_fence(compile_atomic_ordering(fence.ordering()), 0, "");

            None
        }
        Instruction::FreeHeap(free) => {
            builder.build_call(
                instruction_function_set.free_function,
                &[builder
                    .build_bitcast(
                        compile_expression(free.pointer()),
                        context.i8_type().ptr_type(types::DEFAULT_ADDRESS_SPACE),
                        "",
                    )
                    .into()],
                "",
            );

            None
        }
        Instruction::If(if_) => {
            let current = builder.get_insert_block().unwrap();
            let function = current.get_parent().unwrap();

            let then = context.append_basic_block(function, "then");
            let else_ = context.append_basic_block(function, "else");
            let phi = context.append_basic_block(function, "phi");

            let mut cases = vec![];

            builder.build_conditional_branch(
                compile_expression(if_.condition()).into_int_value(),
                then,
                else_,
            );

            for (llvm_block, block) in &[(then, if_.then()), (else_, if_.else_())] {
                builder.position_at_end(*llvm_block);

                let value = compile_block(
                    builder,
                    block,
                    Some(phi),
                    variables,
                    context,
                    target_data,
                    instruction_function_set,
                )?;

                if let Some(value) = value {
                    cases.push((value, builder.get_insert_block().unwrap()));
                }
            }

            builder.position_at_end(phi);

            if cases.is_empty() {
                None
            } else {
                let phi = builder.build_phi(compile_type(if_.type_()), if_.name());

                phi.add_incoming(
                    &cases
                        .iter()
                        .map(|(value, block)| (value as &dyn inkwell::values::BasicValue, *block))
                        .collect::<Vec<_>>(),
                );

                Some(phi.as_basic_value())
            }
        }
        Instruction::Load(load) => Some(builder.build_load(
            compile_expression(load.pointer()).into_pointer_value(),
            load.name(),
        )),
        Instruction::ReallocateHeap(reallocate) => builder
            .build_call(
                instruction_function_set.reallocate_function,
                &[
                    compile_expression(reallocate.pointer()).into(),
                    compile_expression(reallocate.size()).into(),
                ],
                reallocate.name(),
            )
            .try_as_basic_value()
            .left(),
        Instruction::Store(store) => {
            builder.build_store(
                compile_expression(store.pointer()).into_pointer_value(),
                compile_expression(store.value()),
            );

            None
        }
    })
}

fn compile_terminal_instruction<'c>(
    builder: &inkwell::builder::Builder<'c>,
    instruction: &TerminalInstruction,
    destination: Option<inkwell::basic_block::BasicBlock<'c>>,
    variables: &hamt::Map<String, inkwell::values::BasicValueEnum<'c>>,
    context: &'c inkwell::context::Context,
    target_data: &inkwell::targets::TargetData,
    instruction_function_set: &InstructionFunctionSet<'c>,
) -> Option<inkwell::values::BasicValueEnum<'c>> {
    let compile_expression =
        |expression| compile_expression(builder, expression, variables, context, target_data);

    match instruction {
        TerminalInstruction::Branch(branch) => {
            let value = compile_expression(branch.expression());

            builder.build_unconditional_branch(destination.unwrap());

            Some(value)
        }
        TerminalInstruction::Return(return_) => {
            builder.build_return(Some(&compile_expression(return_.expression())));
            None
        }
        TerminalInstruction::Unreachable => {
            if let Some(function) = instruction_function_set.unreachable_function {
                builder.build_call(function, &[], "");
            }

            builder.build_unreachable();

            None
        }
    }
}

fn compile_atomic_ordering(ordering: AtomicOrdering) -> inkwell::AtomicOrdering {
    match ordering {
        AtomicOrdering::Relaxed => inkwell::AtomicOrdering::Monotonic,
        AtomicOrdering::Release => inkwell::AtomicOrdering::Release,
        AtomicOrdering::Acquire => inkwell::AtomicOrdering::Acquire,
        AtomicOrdering::AcquireRelease => inkwell::AtomicOrdering::AcquireRelease,
        AtomicOrdering::SequentiallyConsistent => inkwell::AtomicOrdering::SequentiallyConsistent,
    }
}
