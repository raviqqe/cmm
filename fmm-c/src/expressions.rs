use super::types::*;
use crate::names::*;
use fmm::ir::*;
use fmm::types;

pub fn compile_expression(expression: &Expression) -> String {
    match expression {
        Expression::Primitive(primitive) => compile_primitive(*primitive),
        Expression::Record(record) => {
            format!(
                "({}){{{}}}",
                compile_record_type_id(record.type_()),
                record
                    .elements()
                    .iter()
                    .map(compile_expression)
                    .collect::<Vec<_>>()
                    .join(",")
            )
        }
        Expression::Undefined(undefined) => compile_undefined(undefined),
        Expression::Union(union) => {
            format!(
                "({}){{.{}={}}}",
                compile_union_type_id(union.type_()),
                generate_union_member_name(union.member_index()),
                compile_expression(union.member())
            )
        }
        Expression::Variable(variable) => variable.name().into(),
    }
}

fn compile_undefined(undefined: &Undefined) -> String {
    match undefined.type_() {
        types::Type::Function(_) => "NULL".into(),
        types::Type::Primitive(primitive) => compile_undefined_primitive(*primitive).into(),
        types::Type::Pointer(_) => "NULL".into(),
        types::Type::Record(record) => format!("({}){{}}", compile_record_type_id(record)),
        types::Type::Union(union) => format!("({}){{}}", compile_union_type_id(union)),
    }
}

fn compile_primitive(primitive: Primitive) -> String {
    match primitive {
        Primitive::Boolean(boolean) => format!("{}", boolean),
        Primitive::Float32(number) => format!("{}", number),
        Primitive::Float64(number) => format!("{}", number),
        Primitive::Integer8(number) => format!("{}", number),
        Primitive::Integer32(number) => format!("{}", number),
        Primitive::Integer64(number) => format!("{}", number),
        Primitive::PointerInteger(number) => format!("{}", number),
    }
}

fn compile_undefined_primitive(primitive: types::Primitive) -> &'static str {
    match primitive {
        types::Primitive::Boolean => "false",
        types::Primitive::Float32 | types::Primitive::Float64 => "0.0",
        types::Primitive::Integer8
        | types::Primitive::Integer32
        | types::Primitive::Integer64
        | types::Primitive::PointerInteger => "0",
    }
}
