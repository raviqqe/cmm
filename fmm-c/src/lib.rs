mod error;
mod expressions;
mod instructions;
mod names;
mod types;

pub use error::*;
use expressions::*;
use fmm::analysis::*;
use fmm::ir::*;
use indoc::indoc;
use instructions::*;
use names::*;

const INCLUDES: &str = indoc! {"
    #include <stdatomic.h>
    #include <stdbool.h>
    #include <stdint.h>
    #include <stdlib.h>
"};

pub fn compile(module: &Module) -> String {
    let types = collect_types(module);

    let strings = vec![INCLUDES.into()]
        .into_iter()
        .chain(
            collect_record_types(&types)
                .iter()
                .map(compile_record_type_definition),
        )
        .chain(
            collect_union_types(&types)
                .iter()
                .map(compile_union_type_definition),
        )
        .chain(
            module
                .variable_declarations()
                .iter()
                .map(compile_variable_declaration),
        )
        .chain(
            module
                .variable_definitions()
                .iter()
                .map(compile_variable_forward_declaration),
        )
        .chain(
            module
                .function_declarations()
                .iter()
                .map(compile_function_declaration),
        )
        .chain(
            module
                .function_definitions()
                .iter()
                .map(compile_function_forward_declaration),
        )
        .chain(
            module
                .variable_definitions()
                .iter()
                .map(compile_variable_definition),
        )
        .chain(
            module
                .function_definitions()
                .iter()
                .map(compile_function_definition),
        )
        .collect::<Vec<_>>();

    strings
        .iter()
        .map(|string| string.as_str())
        .collect::<Vec<_>>()
        .join("\n")
}

fn compile_record_type_definition(record: &fmm::types::Record) -> String {
    format!(
        "struct {} {{{}}};",
        generate_record_type_name(record),
        types::compile_record_elements(record)
    )
}

fn compile_union_type_definition(union: &fmm::types::Union) -> String {
    format!(
        "union {} {{{}}};",
        generate_union_type_name(union),
        types::compile_union_members(union)
    )
}

fn compile_variable_declaration(declaration: &VariableDeclaration) -> String {
    "extern ".to_owned() + &types::compile_typed_name(declaration.type_(), declaration.name()) + ";"
}

fn compile_variable_forward_declaration(definition: &VariableDefinition) -> String {
    types::compile_typed_name(
        definition.type_(),
        &(if definition.is_mutable() { "" } else { "const" }.to_owned() + " " + definition.name()),
    ) + ";"
}

fn compile_function_declaration(declaration: &FunctionDeclaration) -> String {
    "extern ".to_owned()
        + &types::compile_function_name(declaration.type_(), declaration.name())
        + ";"
}

fn compile_function_forward_declaration(definition: &FunctionDefinition) -> String {
    types::compile_function_name(definition.type_(), definition.name()) + ";"
}

fn compile_variable_definition(definition: &VariableDefinition) -> String {
    types::compile_typed_name(
        definition.type_(),
        &(if definition.is_mutable() {
            ""
        } else {
            "const "
        }
        .to_owned()
            + definition.name()),
    ) + " = "
        + &compile_expression(definition.body())
        + ";"
}

fn compile_function_definition(definition: &FunctionDefinition) -> String {
    types::compile_function_name(definition.type_(), definition.name())
        + "{"
        + &compile_block(definition.body())
        + "}"
}

#[cfg(test)]
mod tests {
    use super::*;
    use fmm::types;

    #[test]
    fn compile_empty_module() {
        insta::assert_snapshot!(compile(&Module::new(vec![], vec![], vec![], vec![])));
    }

    mod variable_declarations {
        use super::*;

        #[test]
        fn compile_pointer_integer() {
            insta::assert_snapshot!(compile(&Module::new(
                vec![VariableDeclaration::new(
                    "x",
                    types::Primitive::PointerInteger
                )],
                vec![],
                vec![],
                vec![]
            )));
        }

        #[test]
        fn compile_pointer_integer_pointer() {
            insta::assert_snapshot!(compile(&Module::new(
                vec![VariableDeclaration::new(
                    "x",
                    types::Pointer::new(types::Primitive::PointerInteger)
                )],
                vec![],
                vec![],
                vec![]
            )));
        }

        #[test]
        fn compile_function_pointer() {
            insta::assert_snapshot!(compile(&Module::new(
                vec![VariableDeclaration::new(
                    "x",
                    types::Function::new(vec![], types::Primitive::PointerInteger)
                )],
                vec![],
                vec![],
                vec![]
            )));
        }
    }

    mod function_declarations {
        use super::*;

        #[test]
        fn compile_function_pointer() {
            insta::assert_snapshot!(compile(&Module::new(
                vec![],
                vec![FunctionDeclaration::new(
                    "x",
                    types::Function::new(vec![], types::Primitive::PointerInteger)
                )],
                vec![],
                vec![]
            )));
        }
    }

    mod type_definitions {
        use super::*;

        #[test]
        fn compile_record_type_definition() {
            insta::assert_snapshot!(compile(&Module::new(
                vec![VariableDeclaration::new(
                    "x",
                    types::Record::new(vec![types::Primitive::PointerInteger.into()])
                )],
                vec![],
                vec![],
                vec![]
            )));
        }

        #[test]
        fn compile_union_type_definition() {
            insta::assert_snapshot!(compile(&Module::new(
                vec![VariableDeclaration::new(
                    "x",
                    types::Union::new(vec![types::Primitive::PointerInteger.into()])
                )],
                vec![],
                vec![],
                vec![]
            )));
        }
    }
}
