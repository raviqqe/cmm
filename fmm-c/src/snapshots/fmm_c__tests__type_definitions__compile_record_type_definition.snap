---
source: fmm-c/src/lib.rs
expression: "compile(&Module::new(vec![VariableDeclaration ::\n                          new(\"x\", types :: Record ::\n                              new(vec !\n                                  [types :: Primitive :: PointerInteger .\n                                   into()]))], vec![], vec![], vec![]))"
---
#include <stdatomic.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>
struct r_8401336768429464885 {size_t e0;};
extern struct r_8401336768429464885 x;
